pkg/portal/cluster/machines_test.go

```diff
@@ -97,3 +111,154 @@ func TestMachines(t *testing.T) {
+	mockVirtualMachinesClient *mock_compute.MockVirtualMachinesClient
+}
+
+func newMockResourceFactory(mockResourcesClient *mock_features.MockResourcesClient, mockVirtualMachinesClient *mock_compute.MockVirtualMachinesClient) mockResourceFactory {
+	return mockResourceFactory{mockResourcesClient: mockResourcesClient, mockVirtualMachinesClient: mockVirtualMachinesClient}
+}
+
+func (mrf mockResourceFactory) NewResourcesClient(environment *azureclient.AROEnvironment, subscriptionID string, authorizer autorest.Authorizer) features.ResourcesClient {
+	return mrf.mockResourcesClient
+}
+
+func (mrf mockResourceFactory) NewVirtualMachinesClient(environment *azureclient.AROEnvironment, subscriptionID string, authorizer autorest.Authorizer) compute.VirtualMachinesClient {
+	return mrf.mockVirtualMachinesClient
+}
+
+func TestVMAllocationStatus(t *testing.T) {
+	ctx := context.Background()
+	controller := gomock.NewController(t)
+	mockResourcesClient := mock_features.NewMockResourcesClient(controller)
+	mockVirtualMachinesClient := mock_compute.NewMockVirtualMachinesClient(controller)
+	type test struct {
+		name    string
+		mocks   func(*test, *mock_env.MockInterface, *mock_refreshable.MockAuthorizer, *mock_features.MockResourcesClient, *mock_compute.MockVirtualMachinesClient)
+		wantErr string
+	}
+	for _, tt := range []*test{
+		{
+			name: "Successfully fetching VMs allocation status. Calling all the required methods.",
+			mocks: func(tt *test,
+				env *mock_env.MockInterface,
+				authorizer *mock_refreshable.MockAuthorizer,
+				mockResourcesClient *mock_features.MockResourcesClient,
+				mockVirtualMachinesClient *mock_compute.MockVirtualMachinesClient) {
+				env.EXPECT().Environment().Return(&azureclient.AROEnvironment{
+					Environment: azure.Environment{
+						ResourceManagerEndpoint: "temp",
+					},
+				}).AnyTimes()
+				env.EXPECT().FPAuthorizer(gomock.Any(), gomock.Any()).Return(authorizer, nil)
+
+				mockResourcesClient.EXPECT().ListByResourceGroup(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).
+					Return([]mgmtfeatures.GenericResourceExpanded{
+						{
+							Kind: func(v string) *string { return &v }("something"),
+							Type: func(v string) *string { return &v }("Microsoft.Compute/virtualMachines"),
+							Name: func(v string) *string { return &v }("master-x"),
+						},
+					}, nil)
+
+				mockVirtualMachinesClient.EXPECT().Get(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).Return(mgmtcompute.VirtualMachine{
+					Name: func(v string) *string { return &v }("master-x"),
+					VirtualMachineProperties: &mgmtcompute.VirtualMachineProperties{
+						InstanceView: &mgmtcompute.VirtualMachineInstanceView{
+							Statuses: &[]mgmtcompute.InstanceViewStatus{
+								{
+									Code: func() *string {
+										s := new(string)
+										*s = "PowerState/running"
+										return s
+									}(),
+								},
+							},
+						},
+					},
+				}, nil)
+			},
```

Comment:
setting up the mock is better inside the test itself, not in the test data. if it proves difficult to generalize your test, that probably means that it should be in different test functions

Reply:
I should depend on the test case right .. the data we want and test our function according to that data. That should come in the test data only.

Reply:
Sorry, I might have misunderstood you. 
So what I think is, it should depend on the test, support if the test is about a situation where the function will return an error in the first few lines itself, so we would not expect the later function/methods to get executed and we won't write <mockInterface>.EXPECT().<function>.whatever. 

If you meant something else than what I understood, please do elaborate a little.

Reply:
but you should do that setup in the test itself, if there is no expected call you should know it in the data, and set it up accordingly.

Reply:
I have seen this as standard structure in our codebase only .. check this - https://github.com/Azure/ARO-RP/blob/782894071cf91e8c1ca6e182417b851925f823cc/pkg/api/validate/dynamic/dynamic_test.go#L43-L110

Reply:
I know, jim and mj loved that , but it creates loads of code duplication, unreadable and overly complicated tests and we should stop using it

Reply:
I have updated this in the latest push, please do check and let me know.